extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+                                                            )
+                                )
+    );
+
+    /**
+     * Topic-Queue mapping.
+     */
+    final ConcurrentHashMap<String, Set<MessageQueue>> topicQueueMap;
+
+    /**
+     * Default Pull-style client for RocketMQ.
+     */
+    final DefaultMQPullConsumer defaultMQPullConsumer;
+    final DruidPullMessageService pullMessageService;
+
+    messageQueueTreeSetMap.clear();
+    windows.clear();
+
+    try {
+      defaultMQPullConsumer = new DefaultMQPullConsumer(this.consumerGroup);
+      defaultMQPullConsumer.setMessageModel(MessageModel.CLUSTERING);
+      topicQueueMap = new ConcurrentHashMap<>();
+
+      pullMessageService = new DruidPullMessageService(defaultMQPullConsumer);
+      for (String topic : feed) {
+        Validators.checkTopic(topic);
+        topicQueueMap.put(topic, defaultMQPullConsumer.fetchSubscribeMessageQueues(topic));
+      }
+      DruidMessageQueueListener druidMessageQueueListener =
+          new DruidMessageQueueListener(Sets.newHashSet(feed), topicQueueMap, defaultMQPullConsumer);
+      defaultMQPullConsumer.setMessageQueueListener(druidMessageQueueListener);
+      defaultMQPullConsumer.start();
+      pullMessageService.start();
+    }
+    catch (MQClientException e) {
+      LOGGER.error("Failed to start DefaultMQPullConsumer", e);
+      throw new IOException("Failed to start RocketMQ client", e);
+    }
+
+    return new Firehose()
+    {
+
+      @Override
+      public boolean hasMore()
+      {
+        boolean hasMore = false;
+        DruidPullRequest earliestPullRequest = null;
+
+        for (Map.Entry<String, Set<MessageQueue>> entry : topicQueueMap.entrySet()) {
+          for (MessageQueue messageQueue : entry.getValue()) {
+            if (messageQueueTreeSetMap.keySet().contains(messageQueue)
+                && !messageQueueTreeSetMap.get(messageQueue).isEmpty()) {
+              hasMore = true;
+            } else {
+              try {
+                long offset = defaultMQPullConsumer.fetchConsumeOffset(messageQueue, false);
+                int batchSize = (null == pullBatchSize || pullBatchSize.isEmpty()) ?
+                                DEFAULT_PULL_BATCH_SIZE : Integer.parseInt(pullBatchSize);
+
+                DruidPullRequest newPullRequest = new DruidPullRequest(messageQueue, null, offset,
+                                                                       batchSize, !hasMessagesPending()
+                );
+
+                // notify pull message service to pull messages from brokers.
+                pullMessageService.putRequest(newPullRequest);
+
+                // set the earliest pull in case we need to block.
+                if (null == earliestPullRequest) {
+                  earliestPullRequest = newPullRequest;
+                }
+              }
+              catch (MQClientException e) {
+                LOGGER.error("Failed to fetch consume offset for queue: {}", entry.getKey());
```

Comment:
Yes, in case something is wrong. These info are pretty helpful to figure out what's going on, where the message consuming moves to. Anyway, if the caller of this connector have no chance logging these info down, we'd better do this.